File: src/main/java/com/rtsmod/network/payload/UnitUpdatePacket.java
package com.rtsmod.network.payload;

import net.minecraft.network.RegistryByteBuf;
import net.minecraft.network.codec.PacketCodec;
import net.minecraft.network.codec.PacketCodecs;
import net.minecraft.network.packet.CustomPayload;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.BlockPos;

import java.util.Map;
import java.util.UUID;

/**
 * Payload for updating unit positions and states on the client.
 */
public record UnitUpdatePacket(Map<UUID, BlockPos> unitPositions) implements CustomPayload {
    public static final Id<UnitUpdatePacket> ID = new Id<>(Identifier.of("rtsmod", "unit_update"));

    public static final PacketCodec<RegistryByteBuf, UnitUpdatePacket> CODEC = PacketCodec.tuple(
        PacketCodecs.map(java.util.HashMap::new, PacketCodecs.UUID, BlockPos.PACKET_CODEC), UnitUpdatePacket::unitPositions,
        UnitUpdatePacket::new
    );

    @Override
    public Id<? extends CustomPayload> getId() {
        return ID;
    }
}

File: src/main/kotlin/com/rtsmod/network/packets/UnitUpdatePacket.kt
package com.rtsmod.network.packets

import com.rtsmod.core.Unit
import net.minecraft.network.PacketByteBuf
import net.minecraft.util.Identifier

/**
 * Packet for updating unit positions and states
 */
data class UnitUpdatePacket(
    val unitId: String,
    val positionX: Int,
    val positionY: Int,
    val positionZ: Int,
    val health: Int,
    val maxHealth: Int,
    val state: Unit.UnitState,
    val targetX: Int? = null,
    val targetY: Int? = null,
    val targetZ: Int? = null
) {
    
    /**
     * Write packet data to buffer
     */
    fun write(buf: PacketByteBuf) {
        buf.writeString(unitId)
        buf.writeInt(positionX)
        buf.writeInt(positionY)
        buf.writeInt(positionZ)
        buf.writeInt(health)
        buf.writeInt(maxHealth)
        buf.writeString(state.name)
        
        // Write target position if exists
        if (targetX != null && targetY != null && targetZ != null) {
            buf.writeBoolean(true)
            buf.writeInt(targetX)
            buf.writeInt(targetY)
            buf.writeInt(targetZ)
        } else {
            buf.writeBoolean(false)
        }
    }

    companion object {
        /**
         * Read packet data from buffer
         */
        fun read(buf: PacketByteBuf): UnitUpdatePacket {
            val unitId = buf.readString()
            val positionX = buf.readInt()
            val positionY = buf.readInt()
            val positionZ = buf.readInt()
            val health = buf.readInt()
            val maxHealth = buf.readInt()
            val state = Unit.UnitState.valueOf(buf.readString())
            
            val targetX: Int?
            val targetY: Int?
            val targetZ: Int?
            
            if (buf.readBoolean()) {
                targetX = buf.readInt()
                targetY = buf.readInt()
                targetZ = buf.readInt()
            } else {
                targetX = null
                targetY = null
                targetZ = null
            }
            
            return UnitUpdatePacket(
                unitId = unitId,
                positionX = positionX,
                positionY = positionY,
                positionZ = positionZ,
                health = health,
                maxHealth = maxHealth,
                state = state,
                targetX = targetX,
                targetY = targetY,
                targetZ = targetZ
            )
        }
    }
}